File: leetcode/design/leetcode_284.java
package leetcode.design;

import java.util.Arrays;
import java.util.Iterator;
import java.util.List;

public class leetcode_284 {
    public static void main(String[] args) {
        List<Object> peekingList = Arrays.asList(1, 2, 3);
        PeekingIterator peekingIterator = new PeekingIterator(peekingList.iterator());
        System.out.println(peekingIterator.next());
        System.out.println(peekingIterator.peek());
        System.out.println(peekingIterator.next());
        System.out.println(peekingIterator.next());
        System.out.println(peekingIterator.hasNext());
    }

    static class PeekingIterator implements Iterator<Object> {

        Iterator<Object> iterator;
        Object next;

        public PeekingIterator(Iterator<Object> iterator) {
            this.iterator = iterator;
            if (iterator.hasNext())
                this.next = iterator.next();
        }

        public Object peek() {
            return next;
        }

        @Override
        public Object next() {
            Object tmp = next;

            if (iterator.hasNext()) {
                next = iterator.next();
            } else {
                next = null;
            }

            return tmp;
        }

        @Override
        public boolean hasNext() {
            return next != null;
        }
    }
}

File: design/LeetCode_284.kt
package design

class LeetCode_284 {

    class PeekingIterator(
        private val iterator: Iterator<Int>
    ) : Iterator<Int> {

        private var top = 0
        private var isPeeked = false

        fun peek(): Int {
            if (!isPeeked) {
                this.top = this.iterator.next()
                this.isPeeked = true
            }

            return this.top
        }

        override fun next(): Int {
            return if (isPeeked) {
                isPeeked = false
                this.top
            } else {
                this.iterator.next()
            }
        }

        override fun hasNext(): Boolean {
            return isPeeked || this.iterator.hasNext()
        }
    }
}

fun main() {
    // Input
    val iterator = iterator<Int> { 1..3 }
    val peekingIterator = LeetCode_284.PeekingIterator(iterator)
    // Output
    assert(1 == peekingIterator.next())
    assert(2 == peekingIterator.peek())
    assert(2 == peekingIterator.next())
    assert(3 == peekingIterator.next())
    assert(!peekingIterator.hasNext())
}
